File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/calendar/HomeCalendarCreateMenu.kt
@file:Suppress("PackageNaming", "MagicNumber")

package app.pantopus.android.ui.screens.homes.calendar

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * The four create-actions surfaced from the Home calendar FAB sheet.
 * Mirrors iOS `HomeCreateAction`. Only [AddEvent] is in-stream (F3);
 * the others deep-link to other Home streams by route.
 */
enum class HomeCreateAction(
    val tagSuffix: String,
    val icon: PantopusIcon,
    val title: String,
    val subtitle: String,
) {
    AddEvent("addEvent", PantopusIcon.CalendarPlus, "Add event", "A one-off or repeating event"),
    FindATime("findATime", PantopusIcon.Users, "Find a time", "Pick a slot that works for everyone"),
    BookResource("bookResource", PantopusIcon.Package, "Book a resource", "Guest room, EV charger, tools"),
    ScheduleVisit("scheduleVisit", PantopusIcon.DoorOpen, "Schedule a visit", "Offer a vendor or guest a window"),
}

/**
 * Content for the FAB create-menu bottom sheet. The caller hosts it in a
 * `ModalBottomSheet` and dismisses on selection. Mirrors iOS
 * `HomeCalendarCreateMenu`.
 */
@Composable
fun HomeCreateMenuContent(
    onSelect: (HomeCreateAction) -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth().padding(start = Spacing.s2, end = Spacing.s2, bottom = Spacing.s4),
    ) {
        Text(
            text = "Create",
            fontSize = 13.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.appText,
            modifier = Modifier.padding(horizontal = Spacing.s2, vertical = Spacing.s2),
        )
        HomeCreateAction.entries.forEach { action ->
            CreateMenuRow(action = action, onClick = { onSelect(action) })
        }
    }
}

@Composable
private fun CreateMenuRow(
    action: HomeCreateAction,
    onClick: () -> Unit,
) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .clickable(onClick = onClick)
                .padding(horizontal = Spacing.s2, vertical = Spacing.s3)
                .testTag("homeCreateMenu_${action.tagSuffix}"),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Box(
            modifier =
                Modifier
                    .size(38.dp)
                    .clip(RoundedCornerShape(Radii.md))
                    .background(PantopusColors.homeBg),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(icon = action.icon, contentDescription = null, size = 19.dp, tint = PantopusColors.home)
        }
        Column(modifier = Modifier.weight(1f)) {
            Text(text = action.title, fontSize = 14.sp, fontWeight = FontWeight.Bold, color = PantopusColors.appText)
            Text(text = action.subtitle, fontSize = 11.sp, color = PantopusColors.appTextSecondary)
        }
        PantopusIconImage(
            icon = PantopusIcon.ChevronRight,
            contentDescription = null,
            size = Radii.xl,
            tint = PantopusColors.appTextMuted,
        )
    }
}

File: frontend/apps/ios/Pantopus/Features/Scheduling/Home/Calendar/HomeCalendarCreateMenu.swift
//
//  HomeCalendarCreateMenu.swift
//  Pantopus
//
//  Stream I10 — the Home Calendar FAB "create" menu. A bottom sheet that fans
//  out into Add event · Find a time · Book a resource · Schedule a visit.
//  Lifted from `home-calendar-frames.jsx` (`FrameCreateMenu`).
//

import SwiftUI

/// One create-menu action. "Add event" stays in this stream; the rest fan out
/// to other streams' screens (presented locally through the router).
public enum HomeCreateAction: CaseIterable, Hashable {
    case addEvent
    case findATime
    case bookResource
    case scheduleVisit

    var icon: PantopusIcon {
        switch self {
        case .addEvent: .calendarPlus
        case .findATime: .users
        case .bookResource: .package
        case .scheduleVisit: .doorOpen
        }
    }

    var title: String {
        switch self {
        case .addEvent: "Add event"
        case .findATime: "Find a time"
        case .bookResource: "Book a resource"
        case .scheduleVisit: "Schedule a visit"
        }
    }

    var subtitle: String {
        switch self {
        case .addEvent: "A one-off or repeating event"
        case .findATime: "Pick a slot that works for everyone"
        case .bookResource: "Guest room, EV charger, tools"
        case .scheduleVisit: "Offer a vendor or guest a window"
        }
    }
}

/// The create-menu sheet content (presented via `.sheet`).
struct HomeCalendarCreateMenu: View {
    let onSelect: @MainActor (HomeCreateAction) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s0) {
            Text("Create")
                .font(.system(size: 13, weight: .bold))
                .foregroundStyle(Theme.Color.appText)
                .padding(.horizontal, Spacing.s2)
                .padding(.top, Spacing.s2)
                .padding(.bottom, Spacing.s2)
            ForEach(HomeCreateAction.allCases, id: \.self) { action in
                Button {
                    onSelect(action)
                } label: {
                    HStack(spacing: Spacing.s3) {
                        Icon(action.icon, size: 19, color: Theme.Color.home)
                            .frame(width: 38, height: 38)
                            .background(Theme.Color.homeBg)
                            .clipShape(RoundedRectangle(cornerRadius: Radii.md, style: .continuous))
                        VStack(alignment: .leading, spacing: 1) {
                            Text(action.title)
                                .font(.system(size: 14, weight: .bold))
                                .foregroundStyle(Theme.Color.appText)
                            Text(action.subtitle)
                                .font(.system(size: 11))
                                .foregroundStyle(Theme.Color.appTextSecondary)
                        }
                        Spacer(minLength: 0)
                        Icon(.chevronRight, size: 16, color: Theme.Color.appTextMuted)
                    }
                    .padding(.horizontal, Spacing.s2)
                    .padding(.vertical, Spacing.s3)
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("homeCreateMenu_\(action)")
            }
        }
        .padding(.horizontal, Spacing.s2)
        .padding(.bottom, Spacing.s4)
        .presentationDetents([.height(340)])
        .presentationDragIndicator(.visible)
    }
}
